File: Data Structure Final Project/BinaryTree.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Data_Structure_Final_Project
{
    internal class BinaryTree
    {
        public class Node
        {
            public Node? left;
            public Node? right;
            public int? value;

            public Node(int value)
            {
                this.value = value;
                left = null;
                right = null;

            }
            public Node()
            {
                value = null;
                left = null;
                right = null;

            }

            public void Insert(int value)
            {
                if (this.value == null)
                {
                    this.value = value;
                    return;
                }
                if (value < this.value)
                {
                    if (left == null)
                    {
                        left = new Node(value);

                    }
                    else
                    {
                        left.Insert(value);
                    }

                }
                else
                {
                    if (right == null)
                    {
                        right = new Node(value);

                    }
                    else
                    {
                        right.Insert(value);
                    }
                }




            }




            public Node? Search(int value)
            {
                if (this.value == null) return null;

                if (this.value == value) return this;

                if (value < this.value)
                {
                    if (left != null) return left.Search(value);
                    else return null;
                }
                else
                {
                    if (right != null) return right.Search(value);
                    else return null;
                }
            }

            public Node? Delete(Node? root, int value)
            {
                if (root == null) return null;

                if (value < root.value)
                {
                    root.left = Delete(root.left, value);
                }
                else if (value > root.value)
                {
                    root.right = Delete(root.right, value);
                }
                else
                {

                    if (root.left == null && root.right == null)
                    {
                        return null;
                    }
                    else if (root.left == null)
                    {
                        return root.right;
                    }
                    else if (root.right == null)
                    {
                        return root.left;
                    }
                    else
                    {
                        Node? minNode = FindMin(root.right);
                        root.value = minNode.value;
                        root.right = Delete(root.right, (int)minNode.value);
                    }
                }

                return root;
            }

            private Node? FindMin(Node? node)
            {
                while (node?.left != null)
                {
                    node = node.left;
                }
                return node;
            }

            public string GetNodeInfo(Node? node, string action)
            {


                if (node == null)
                {
                    return "Node with value not found.";
                }

                int height = GetNodeHeight(node);



                bool isLeaf = node.left == null && node.right == null;

                string leafStatus = isLeaf ? "is a leaf node" : "is not a leaf node";

                return $"The {action} Node with value {node.value.ToString()} was found at the height {height} and {leafStatus}.";
            }
            public int GetNodeHeight(Node? node)
            {

                if (this == null) return 0;
                int height = 1;
                if (value == node?.value)
                    return height;

                if (node?.value < value)
                {


                    if (left != null)
                    {

                        if (left.value == node?.value)
                        {
                            return height + 1;
                        }
                        int temp = left.GetNodeHeight(node);
                        return height + temp;
                    }
                    else return 0;
                }
                else
                {
                    if (right != null)
                    {


                        if (right.value == node?.value)
                        {
                            return height + 1;
                        }
                        return height + right.GetNodeHeight(node);

                    }
                    else return 0;
                }
            }


            public int CountNodes()
            {
                int count = 1;
                if (left != null)
                {
                    count += left.CountNodes();
                }
                if (right != null)
                {
                    count += right.CountNodes();
                }
                return count;
            }

            public int GetHeight()
            {
                if (this == null) return 0;

                int leftHeight = left?.GetHeight() ?? 0;
                int rightHeight = right?.GetHeight() ?? 0;

                return Math.Max(leftHeight, rightHeight) + 1;
            }




            public void TreeInfo()
            {
                int heightt = GetHeight();
                int nodeCount = CountNodes();
                int height = heightt;
                int levels = height;

                MessageBox.Show($"Tree Information:\nNodes: {nodeCount}\nHeight: {height}\nLevels: {levels}");
            }
        }
    }
}

File: Data Structure Final Project/BinaryTree.vb
Imports System

Namespace Data_Structure_Final_Project
    Friend Class BinaryTree
        Public Class Node
            Public left As Node?
            Public right As Node?
            Public value As Integer?

            Public Sub New(value As Integer)
                Me.value = value
                left = Nothing
                right = Nothing

            End Sub
            Public Sub New()
                value = Nothing
                left = Nothing
                right = Nothing

            End Sub

            Public Sub Insert(value As Integer)
                If Me.value Is Nothing Then
                    Me.value = value
                    Return
                End If
                If value < Me.value Then
                    If left Is Nothing Then
                        left = New Node(value)
                    Else
                        left.Insert(value)
                    End If
                Else
                    If right Is Nothing Then
                        right = New Node(value)
                    Else
                        right.Insert(value)
                    End If
                End If




            End Sub




            Public Function Search(value As Integer) As Node?
                If Me.value Is Nothing Then Return Nothing

                If Me.value = value Then Return Me

                If value < Me.value Then
                    If left IsNot Nothing Then
                        Return left.Search(value)
                    Else
                        Return Nothing
                    End If
                Else
                    If right IsNot Nothing Then
                        Return right.Search(value)
                    Else
                        Return Nothing
                    End If
                End If
            End Function

            Public Function Delete(root As Node?, value As Integer) As Node?
                If root Is Nothing Then Return Nothing

                If value < root.value Then
                    root.left = Me.Delete(root.left, value)
                ElseIf value > root.value Then
                    root.right = Me.Delete(root.right, value)
                Else

                    If root.left Is Nothing AndAlso root.right Is Nothing Then
                        Return Nothing
                    ElseIf root.left Is Nothing Then
                        Return root.right
                    ElseIf root.right Is Nothing Then
                        Return root.left
                    Else
                        Dim minNode = Me.FindMin(root.right)
                        root.value = minNode.value
                        root.right = Me.Delete(root.right, CInt(minNode.value))
                    End If
                End If

                Return root
            End Function

            Private Function FindMin(node As Node?) As Node?
                While node?.left IsNot Nothing
                    node = node.left
                End While
                Return node
            End Function

            Public Function GetNodeInfo(node As Node?, action As String) As String


                If node Is Nothing Then
                    Return "Node with value not found."
                End If

                Dim height = GetNodeHeight(node)



                Dim isLeaf = node.left Is Nothing AndAlso node.right Is Nothing

                Dim leafStatus = If(isLeaf, "is a leaf node", "is not a leaf node")

                Return $"The {action} Node with value {node.value.ToString()} was found at the height {height} and {leafStatus}."
            End Function
            Public Function GetNodeHeight(node As Node?) As Integer

                If Me Is Nothing Then Return 0
                Dim height = 1
                If value = node?.value Then Return height

                If node?.value < value Then


                    If left IsNot Nothing Then

                        If left.value = node?.value Then
                            Return height + 1
                        End If
                        Dim temp As Integer = left.GetNodeHeight(node)
                        Return height + temp
                    Else
                        Return 0
                    End If
                Else
                    If right IsNot Nothing Then


                        If right.value = node?.value Then
                            Return height + 1
                        End If
                        Return height + right.GetNodeHeight(node)
                    Else
                        Return 0
                    End If
                End If
            End Function


            Public Function CountNodes() As Integer
                Dim count = 1
                If left IsNot Nothing Then
                    count += left.CountNodes()
                End If
                If right IsNot Nothing Then
                    count += right.CountNodes()
                End If
                Return count
            End Function

            Public Function GetHeight() As Integer
                If Me Is Nothing Then Return 0

                Dim leftHeight As Integer = If(left?.GetHeight(), 0)
                Dim rightHeight As Integer = If(right?.GetHeight(), 0)

                Return Math.Max(leftHeight, rightHeight) + 1
            End Function




            Public Sub TreeInfo()
                Dim heightt As Integer = GetHeight()
                Dim nodeCount As Integer = CountNodes()
                Dim height = heightt
                Dim levels = height

                System.Windows.Forms.MessageBox.Show($"Tree Information:
Nodes: {nodeCount}
Height: {height}
Levels: {levels}")
            End Sub
        End Class
    End Class
End Namespace
